numpy/core/src/umath/ufunc_type_resolution.c

```diff
@@ -1508,6 +1507,22 @@ ufunc_loop_matches(PyUFuncObject *self,
             tmp = dtypes[i];
             Py_INCREF(tmp);
         }
+        else if (types[i] == PyArray_TYPE(op[i])) {
+            /*
+             * Don't bother doing casting checks if the type numbers exactly
+             * match.
```

Comment:
> I think if we truly just want a special case here, it's safer/clearer to just add a new type resolver instead of adding an obscure branch to the default resolver to make one datetime ufunc work. Does that seem reasonable?

Yes, agreed. We *could* make it slightly more generic (e.g., to handle all datetime64->bool signatures instead of just `isfinite`) but this sounds good for now.